### PR TITLE
Optimize recsys-FlexKV

### PR DESCRIPTION
**Pull request by Clebrate**
## Description
- Optimize `_build_slot_mappings ` by reducing D2H datasize.
- Optimize `as_batch`
- support SSD

### Performance improvement summary

1. **`slot_mapping_optimization`** 

#### Before vs. after at `seq_len = 2048` (L1, ms)

| op | Before | After | Δ |
| --- | --- | --- | --- |
| `offload_launch` | 7.192 | 3.345 | **−53.5%** |
| `offload_wait` | 30.655 | 22.422 | **−26.9%** |
| `onboard_launch` | 1.380 | 0.696 | **−49.6%** |
| `onboard_wait` | 21.244 | 19.341 | **−9.0%** |
| offload launch+wait | 37.847 | 25.767 | **−31.9%** |
| onboard launch+wait | 22.624 | 20.037 | **−11.4%** |
| **total pipeline** | 67.377 | 54.085 | **−19.7%** |

**Main gains:**
- **Offload launch**: `_build_slot_mappings` drops from **3.30 ms → 0.43 ms** (−87%).
- **Offload wait**: **−15% ~ −37%** across seq lengths.
- **Onboard launch**: batch `launch` cuts onboard launch **~−50%**.

#### Launch + Wait Summary (`flexkv_as_batch=0` vs `flexkv_as_batch=1`)

| path | seq_len | `as_batch=0` (ms) | `as_batch=1` (ms) | Δ |
| --- | --- | --- | --- | --- |
| offload | 1024 | 26.87 | 18.63 | **−30.7%** |
| offload | 2048 | 30.59 | 25.77 | **−15.7%** |
| offload | 4096 | 39.17 | 38.58 | −1.5% |
| onboard | 1024 | 17.51 | 14.65 | **−16.4%** |
| onboard | 2048 | 26.31 | 20.04 | **−23.9%** |
| onboard | 4096 | 40.10 | 30.59 | **−23.7%** |

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/recsys-examples/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by JacoCheung** [comment]
:x: **Pipeline #55646281** -- failed

| Job | Status | Log |
|-----|--------|-----|
| pre_check | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622421) |
| train_build_x86 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622422) |
| train_build_arm64 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622423) |
| tritonserver_build_x86 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622424) |
| prepare-jet-b200-smoke | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622425) |
| prepare-jet-b200-inference | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622427) |
| prepare-jet-cw-dfw-e2e-benchmark | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622429) |
| build_whl | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622431) |
| dynamicemb_test_fwd_bwd_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622432) |
| dynamicemb_test_load_dump_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622433) |
| unit_test_1gpu_a100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622434) |
| unit_test_1gpu_h100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622435) |
| unit_test_4gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622436) |
| unit_test_tp_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622437) |
| L20_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622438) |
| inference_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622439) |
| inference_test_1gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/346622440) |

**Result: 12/17 jobs passed**

[View full pipeline](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/55646281)

File: corelib/recsys_kvcache_manager/recsys_kvcache_manager/flex_kvcache_manager.py [hunk 1]
```diff
@@ -458,18 +476,55 @@ def offload_kvcache_launch(
...
+                    num_matched = aligned_len - num_unmatched
+                    start = (num_matched // self.page_size) * self.page_size
+                    end = start + (num_unmatched // self.page_size) * self.page_size
+                    if end > start and end <= slot_mapping.numel():
+                        launch_slot = slot_mapping[start:end].contiguous()
+                if launch_slot is None:
+                    self._client.cancel(int(task_id))
+                    continue
+                task_ids.append(int(task_id))
+                batch_slot_mappings.append(launch_slot)
```
[the remark this answers is not in the document]
Comment (Clebrate) [reply corelib/recsys_kvcache_manager/recsys_kvcache_manager/flex_kvcache_manager.py:518]: Skip this suggestion. In our current FlexKV flow, offload/release is page-granular, so a sub-page sequence (`valid_len < page_size`) does not result in page eviction. 
Even if that UID is included in the handle, `release_offload_pages` updates effective offloaded pages using integer page math, and for sub-page lengths this resolves to 0 pages. So no GPU pages are actually released for that short sequence, and no immediate data loss occurs in this path.

File: corelib/recsys_kvcache_manager/configs/flexkv_ssd.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# FlexKV tier config for recsys KVCacheManager.
+
+cpu_cache_gb: 0.2
+ssd_cache_gb: 0.5
+ssd_cache_dir: /home/scratch.noliu_gpu/ssd_cache
```
[the remark this answers is not in the document]
Comment (Clebrate) [reply corelib/recsys_kvcache_manager/configs/flexkv_ssd.yml]: Fixed

File: corelib/recsys_kvcache_manager/recsys_kvcache_manager/flex_kvcache_manager.py [hunk 3]
```diff
@@ -284,25 +299,26 @@ def lookup_kvcache(self, index_meta: KVIndexMeta) -> KVLookupResult:
...
-            if page_ids.numel() == 0:
+        kv_indices_cpu = (
+            kv_indices.detach().contiguous().to(device="cpu", dtype=torch.int64)
+        )
+        kv_indptr_cpu = kv_indptr.detach().contiguous().to(
+            device="cpu", dtype=torch.int64
+        )
+        indptr = kv_indptr_cpu.tolist()
+        offsets = torch.arange(self.page_size, dtype=torch.int64)
+        mappings: List[torch.Tensor] = []
```
[the remark this answers is not in the document]
Comment (Clebrate) [reply corelib/recsys_kvcache_manager/recsys_kvcache_manager/flex_kvcache_manager.py]: Fixed

**Comment by JacoCheung** [comment]
:x: **Pipeline #55763541** -- failed

| Job | Status | Log |
|-----|--------|-----|
| pre_check | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553235) |
| train_build_x86 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553236) |
| train_build_arm64 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553237) |
| tritonserver_build_x86 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553238) |
| prepare-jet-b200-smoke | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553239) |
| prepare-jet-b200-inference | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553241) |
| prepare-jet-cw-dfw-e2e-benchmark | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553243) |
| build_whl | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553245) |
| dynamicemb_test_fwd_bwd_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553246) |
| dynamicemb_test_load_dump_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553247) |
| unit_test_1gpu_a100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553249) |
| unit_test_1gpu_h100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553250) |
| unit_test_4gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553252) |
| unit_test_tp_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553253) |
| L20_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553254) |
| inference_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553255) |
| inference_test_1gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/347553256) |

**Result: 13/17 jobs passed**

[View full pipeline](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/55763541)

**Comment by JacoCheung** [comment]
:x: **Pipeline #57256931** -- failed

| Job | Status | Log |
|-----|--------|-----|
| pre_check | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705013) |
| train_build_x86 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705014) |
| train_build_arm64 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705015) |
| tritonserver_build_x86 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705016) |
| prepare-jet-b200-smoke | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705017) |
| prepare-jet-b200-inference | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705019) |
| prepare-jet-cw-dfw-e2e-benchmark | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705021) |
| build_whl | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705023) |
| dynamicemb_test_fwd_bwd_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705024) |
| dynamicemb_test_load_dump_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705025) |
| unit_test_1gpu_a100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705026) |
| unit_test_1gpu_h100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705027) |
| unit_test_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705028) |
| unit_test_tp_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705029) |
| L20_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705030) |
| inference_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705031) |
| inference_test_1gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/357705032) |

**Result: 11/17 jobs passed**

[View full pipeline](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/57256931)

File: corelib/recsys_kvcache_manager/test/FLEXKV_CPU_BREAKDOWN.md [hunk 4]
```diff
@@ -15,13 +16,13 @@
 | `num_heads` | **4** |
 | `batch_size` | **8** |
 | `len_per_seq` / `sequence_length` | **1024 / 2048 / 4096** |
+| `page_size` | **32** |
 | `dtype` | **bf16** |
 
-
 ```text
```
Comment (shijieliu) [inline corelib/recsys_kvcache_manager/test/FLEXKV_CPU_BREAKDOWN.md]: maybe remove the text wrapper here?

File: corelib/recsys_kvcache_manager/test/FLEXKV_CPU_BREAKDOWN.md [hunk 5]
```diff
@@ -1,8 +1,9 @@
 # Recsys FlexKVCache Manager — CPU Breakdown Analysis
```
Comment (shijieliu) [inline corelib/recsys_kvcache_manager/test/FLEXKV_CPU_BREAKDOWN.md:1]: could you summarize the perf improvement in the PR?

**Comment by JacoCheung** [comment]
:x: **Pipeline #57502046** -- failed

| Job | Status | Log |
|-----|--------|-----|
| pre_check | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674226) |
| train_build_x86 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674227) |
| train_build_arm64 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674228) |
| tritonserver_build_x86 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674229) |
| prepare-jet-b200-smoke | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674230) |
| prepare-jet-b200-inference | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674232) |
| prepare-jet-cw-dfw-e2e-benchmark | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674234) |
| build_whl | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674236) |
| dynamicemb_test_fwd_bwd_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674237) |
| dynamicemb_test_load_dump_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674238) |
| unit_test_1gpu_a100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674239) |
| unit_test_1gpu_h100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674240) |
| unit_test_4gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674241) |
| unit_test_tp_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674242) |
| L20_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674243) |
| inference_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674244) |
| inference_test_1gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/359674245) |

**Result: 12/17 jobs passed**

[View full pipeline](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/57502046)

**Review by shijieliu — APPROVED** [review]

**Review by shijieliu — APPROVED** [review]
